171648174 Add reset password

Agnes-N opened the pull request:
#### Description
This PR is for resetting the user password when he/she had forgotten it.
#### Type of change

Please select the relevant option

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Others (cosmetics, styling, improvements)
- [ ] This change requires a documentation update

#### How Has This Been Tested?

- [x] Unit
- [ ] Integration
- [ ] End-to-end
- [ ] No testing required

#### How to Test

- Clone the champs-backend repo to your machine
- Navigate the cloned folder and open it 
- In your terminal run this command "npm run serve"
- Then go to the postman and put in this URL "localhost:3000/api/v1/reset-link" and hit the send button

#### Any background context you want to add (Optional)

![end](https://user-images.githubusercontent.com/51400717/78127353-4a1ffe00-7414-11ea-9970-d069b1317916.png) 

#### STORY ID

[Finishes #171648174]

placideirandora commented on server/controllers/resetController.js:
```diff
@@ -0,0 +1,52 @@
+const { User } = models;
+class ResetController {
+
+  /**This method is for to handle the request to signup
```
Consider updating your comment. The method sends a password reset link instead.

Agnes-N replied:
Done

placideirandora commented on server/helpers/authHelpers.js:
```diff
@@ -0,0 +1,60 @@
+   * @returns {object} email of the user
+  */
+
+  static async gotUser(attr, val) {
```
Consider updating the method name to **findUser** instead.

Agnes-N replied:
Done

placideirandora commented on server/tests/passwordResetTest.test.js:
```diff
@@ -0,0 +1,90 @@
+            .post('/api/v1/reset-link')
+            .send({ email: user.email })
+            .end((err, res) => { 
+                expect(res.body.status).to.be.equal(200);
```
Consider testing the message property as well. You should test all properties of the object response.

Agnes-N replied:
Done

placideirandora commented on server/tests/passwordResetTest.test.js:
```diff
@@ -0,0 +1,90 @@
+            .post('/api/v1/reset-link')
+            .send({ email: user.email1 })
+            .end((err, res) => {
+                expect(res.body.status).to.be.equal(404);
```
Consider testing the message property as well. You should test all properties of the object response.

Agnes-N replied:
Done

placideirandora commented on server/tests/passwordResetTest.test.js:
```diff
@@ -0,0 +1,90 @@
+            .post('/api/v1/reset-link')
+            .send({ email: user.email2 })
+            .end((err, res) => {
+                expect(res.body.status).to.be.equal(404);
```
The same applies here.

Agnes-N replied:
Done

placideirandora commented on server/tests/passwordResetTest.test.js:
```diff
@@ -0,0 +1,90 @@
+            .post('/api/v1/reset-link')
+            .send({ email: user.email3 })
+            .end((err, res) => {
+                expect(res.body.status).to.be.equal(404);
```
The same applies here.

Agnes-N replied:
Done

placideirandora commented on server/tests/passwordResetTest.test.js:
```diff
@@ -0,0 +1,90 @@
+        router()
+            .get('/api/v1/res2ink') 
+            .end((err, res) => {
+                expect(res.body.status).to.be.equal(200);
```
The same applies here.

Agnes-N replied:
Done

placideirandora commented on server/tests/passwordResetTest.test.js:
```diff
@@ -0,0 +1,90 @@
+            .get('/update-password/{email}/{token}') 
+            .send({password: 'aggggg'})
+            .end((err, res) => {
+                expect(res.body.status).to.be.equal(200);
```
The same applies here.

Agnes-N replied:
Done

placideirandora commented on server/tests/passwordResetTest.test.js:
```diff
@@ -0,0 +1,90 @@
+            .get('/update-password/{email}/{token}') 
+            .send({password: 'aggggg'})
+            .end((err, res) => {
+                expect(res.body.status).to.be.equal(200);
```
The same applies here.

Agnes-N replied:
Done

Agnes-N committed:
171648174-feat(reset password): Add reset password

- create an endpoint for sending a password reset link
- create an endpoint for resetting password

[Finished #171648174]

Agnes-N commented on server/config/winston.js:
```diff
@@ -0,0 +1,96 @@
+    level: 'http',
+    filename: 'server/logs/http.logs',
+    colorize: false,
+    format: combine(
```
resolved

Agnes-N commented on server/helpers/authHelpers.js:
```diff
@@ -0,0 +1,99 @@
+  */
+  static async createToken(data) {
+    const token = jwt.sign(data,
+      process.env.SECRET_KEY);
```
resolved

Agnes-N commented on server/helpers/authHelpers.js:
```diff
@@ -0,0 +1,99 @@
+   * @returns {object} The status and some data of the user.
+  */
+  static async createToken(data) {
+    const token = jwt.sign(data,
```
resolved

Agnes-N commented on server/config/winston.js:
```diff
@@ -0,0 +1,96 @@
+    colorize: false,
+    format: combine(
+      label({ label: 'Error' }),
+    )
```
resolved

Agnes-N commented on server/config/winston.js:
```diff
@@ -0,0 +1,96 @@
+    level: 'error',
+    filename: 'server/logs/error.logs',
+    colorize: false,
+    format: combine(
```
resolved

Agnes-N commented on server/config/winston.js:
```diff
@@ -0,0 +1,96 @@
+    colorize: false,
+    format: combine(
+      label({ label: 'Info' }),
+    ),
```
resolved

Agnes-N commented on server/config/winston.js:
```diff
@@ -0,0 +1,96 @@
+    level: 'info',
+    filename: 'server/logs/info.logs',
+    colorize: false,
+    format: combine(
```
resolved

Agnes-N commented on server/config/winston.js:
```diff
@@ -0,0 +1,96 @@
+    colorize: false,
+    format: combine(
+      label({ label: 'Http' }),
+    ),
```
resolved

placideirandora commented on server/controllers/resetController.js:
```diff
@@ -0,0 +1,75 @@
+    const { email } = req.body;
+    const registeredUser = await authHelper.findUser(email);
+    const task = 'password-reset';
+
```
Remove this one excessive blank line.

Agnes-N replied:
updated

placideirandora commented on server/controllers/resetController.js:
```diff
@@ -0,0 +1,75 @@
+      await authHelper.sendMail(email, token, task);
+      res.status(200).send({
+        status: 200,
+        message: 'check your email to reset password!'
```
Consider hanging the message to '**Password reset link has been sent to your email. Check it out!**'

Agnes-N replied:
updated

placideirandora commented on server/controllers/resetController.js:
```diff
@@ -0,0 +1,75 @@
+ */
+class ResetController {
+  /**
+     * This method find the user email request.
```
Consider changing the comment to '**This method checks if the user is already registered prior to resetting the password**'.

Agnes-N replied:
updated

placideirandora commented on server/helpers/authHelpers.js:
```diff
@@ -1,44 +1,53 @@
- * required to handle
- * authentication data to the database.
+ * required to handle all
+ * password reset and update.
```
Consider updating this comment to fit the current use. It is not only about the password. Use a comment that is general in this case to cover each use.

Agnes-N replied:
updated

placideirandora replied:
Not yet, Double-check. The comment has not changed. This class is handling several things including sending emails, generating, and verifying tokens. Thus, find an appropriate comment that encompasses the functionalities.

Agnes-N replied:
check again to see if it is okay now

placideirandora commented on server/middlewares/validateAuth.js:
```diff
@@ -48,5 +50,66 @@ class validateAuth {
   }
+
+  /**
+     * This method handle the reset password request.
```
Consider updating this comment as well. It validates instead.

Agnes-N replied:
updated

placideirandora commented on server/middlewares/validateAuth.js:
```diff
@@ -48,5 +50,66 @@ class validateAuth {
+  }
+
+  /**
+       * This method handle the update of reset password request.
```
Update this comment as well in a correct and meaningful way.

Agnes-N replied:
updated

placideirandora commented on server/middlewares/validateAuth.js:
```diff
@@ -48,5 +50,66 @@ class validateAuth {
+  }
+
+  /**
+       * This method handle the update of reset password request.
```
Update this comment as well.

Agnes-N replied:
updated

placideirandora commented on server/middlewares/validePasswordReset.js:
```diff
@@ -0,0 +1,81 @@
+import jwt from 'jsonwebtoken';
```
Delete this file because It is not being used anywhere else. As I remember, We have imported its functionality in the authHelper instead. But double-check to confirm that.

Agnes-N replied:
Resolved
